app/crud/base.py

```diff
@@ -0,0 +1,122 @@
+from snowflake.snowpark import Row
+from pydantic import BaseModel
+from typing import ClassVar, get_args, get_origin, Union, Dict
+import datetime
+
+
+## TODO
+# test validation stuff
+# migrate - maybe we don't move out of sql?
+# copy predefined labels to labels - maybe we dont move out of sql?
+# same again for probes
+# hook up to stored procs for (CRUD and validation)
+# more tests
+class BaseOpsCenterModel(BaseModel):
+    # The name of the table in snowflake (without schema) that the model maps to.
+    table_name: ClassVar[str] = None
+    # The procedure to be called after any CRUD operation on this model.
+    on_success_proc: ClassVar[str] = None
+
+    @classmethod
+    def cols_dict(cls) -> Dict[str, str]:
+        cols = dict()
+        for field, info in cls.__fields__.items():
+            if isinstance(type, type(info.annotation)):
+                info_type = handle_type(info.annotation)
+            else:
+                info_type = handle_union(
+                    get_args(info.annotation), get_origin(info.annotation)
+                )
+            cols[field] = info_type
+        return cols
+
+    @classmethod
+    def create_table(cls, session, with_catalog_view=True):
+        cols = cls.cols_dict()
+        cols_str = ", ".join([f"{k} {v}" for k, v in cols.items()])
+        session.sql(
+            f"CREATE TABLE IF NOT EXISTS internal.{cls.table_name} ({cols_str})"
+        ).collect()
+        if with_catalog_view:
+            session.sql(
+                f"CREATE OR REPLACE VIEW catalog.{cls.table_name} AS SELECT * FROM internal.{cls.table_name}"
+            ).collect()
+
+    def write(self, session):
+        try:
+            session.sql("BEGIN").collect()
+            df = session.create_dataframe([Row(**dict(self))])
+            df.write.mode("append").save_as_table(f"INTERNAL.{self.table_name}")
+            session.sql("COMMIT").collect()
+        except Exception:
+            session.sql("ROLLBACK").collect()
+            raise
+
+        session.call(self.on_success_proc)
+
+    def get_id(self) -> str:
+        return None
```

Comment:
nit: can you add pydoc? I know I wrote it so I feel a bit embarassed asking

Reply:
👍🏼  no problem